extern/sector-storage/piece_provider.go

```diff
@@ -0,0 +1,146 @@
+}
+
+type pieceProvider struct {
+	storage *stores.Remote
+	index   stores.SectorIndex
+	uns     Unsealer
+}
+
+func NewPieceProvider(storage *stores.Remote, index stores.SectorIndex, uns Unsealer) PieceProvider {
+	return &pieceProvider{
+		storage: storage,
+		index:   index,
+		uns:     uns,
+	}
+}
+
+// tryReadUnsealedPiece will try to read the unsealed piece from an existing unsealed sector file for the given sector from any worker that has it.
+// It will NOT try to schedule an Unseal of a sealed sector file for the read.
+//
+// Returns a nil reader if the piece does NOT exist in any unsealed file or there is no unsealed file for the given sector on any of the workers.
+func (p *pieceProvider) tryReadUnsealedPiece(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize) (io.ReadCloser, context.CancelFunc, error) {
+	// acquire a lock purely for reading unsealed sectors
+	ctx, cancel := context.WithCancel(ctx)
+	if err := p.index.StorageLock(ctx, sector.ID, storiface.FTUnsealed, storiface.FTNone); err != nil {
+		cancel()
+		return nil, nil, xerrors.Errorf("acquiring read sector lock: %w", err)
+	}
+
+	// Reader returns a reader for an unsealed piece at the given offset in the given sector.
+	// The returned reader will be nil if none of the workers has an unsealed sector file containing
+	// the unsealed piece.
+	r, err := p.storage.Reader(ctx, sector, abi.PaddedPieceSize(offset.Padded()), size.Padded())
+	if err != nil {
+		log.Debugf("failed storage reader;sector=%+v, err:%s", sector.ID, err)
+		cancel()
+		return nil, nil, err
+	}
+	if r == nil {
+		cancel()
+	}
+
+	return r, cancel, nil
+}
+
+// ReadPiece is used to read an Unsealed piece at the given offset and of the given size from a Sector
+// If an Unsealed sector file exists with the Piece Unsealed in it, we'll use that for the read.
+// Otherwise, we will Unseal a Sealed sector file for the given sector and read the Unsealed piece from it.
+func (p *pieceProvider) ReadPiece(ctx context.Context, sector storage.SectorRef, offset storiface.UnpaddedByteIndex, size abi.UnpaddedPieceSize, ticket abi.SealRandomness, unsealed cid.Cid) (io.ReadCloser, bool, error) {
+	if err := offset.Valid(); err != nil {
+		return nil, false, xerrors.Errorf("offset is not valid: %w", err)
+	}
+	if err := size.Validate(); err != nil {
+		return nil, false, xerrors.Errorf("size is not a valid piece size: %w", err)
+	}
+
+	r, unlock, err := p.tryReadUnsealedPiece(ctx, sector, offset, size)
+
+	log.Infof("tryReadUnsealedPiece result: r=%+v, err=%s", r, err)
+
+	if xerrors.Is(err, storiface.ErrSectorNotFound) {
+		log.Debugf("no unsealed sector file with unsealed piece, sector=%+v, offset=%d, size=%d", sector, offset, size)
+		err = nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	var uns bool
+
+	if r == nil {
```

Comment:
if `r == nil` we don't have to call `unlock` ? Because we do have a few returns in this branch that are not calling `unlock`...?

Reply:
If r == nil unlock is called within tryReadUnsealedPiece:
https://github.com/filecoin-project/lotus/pull/6280/files/40642b2cad0bb247dd789837c6465163e5b59742#diff-1fe5f948c5ef6d7d11919b998fedd74927b0cfac11ebb3ef1dbaf518fc026999R61

I found this very hard to follow as well, I wonder if there's a way we can refactor to simplify